# Cleaning used cars database
## Over 370,000 used cars scraped from Ebay Kleinanzeigen

## Importing packages and the dataset

In [2]:
import pandas as pd
import numpy as np
%matplotlib inline

In [3]:
data = pd.read_csv('Data/autos.csv', encoding='Latin-1')

## Exploring the dataset

In [4]:
data.shape

(371528, 20)

In [5]:
data.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

In [7]:
data.describe()

price  yearOfRegistration        powerPS      kilometer  \
count  3.715280e+05       371528.000000  371528.000000  371528.000000   
mean   1.729514e+04         2004.577997     115.549477  125618.688228   
std    3.587954e+06           92.866598     192.139578   40112.337051   
min    0.000000e+00         1000.000000       0.000000    5000.000000   
25%    1.150000e+03         1999.000000      70.000000  125000.000000   
50%    2.950000e+03         2003.000000     105.000000  150000.000000   
75%    7.200000e+03         2008.000000     150.000000  150000.000000   
max    2.147484e+09         9999.000000   20000.000000  150000.000000   

       monthOfRegistration  nrOfPictures    postalCode  
count        371528.000000      371528.0  371528.00000  
mean              5.734445           0.0   50820.66764  
std               3.712412           0.0   25799.08247  
min               0.000000           0.0    1067.00000  
25%               3.000000           0.0   30459.00000  
50%               6.000000           0.0   49610.00000  
75%               9.000000           0.0   71546.00000  
max              12.000000           0.0   99998.00000

Based on our quick exploration of the dataset we make the following observations:
• Column names are not homogeneous
• A bunch of columns have missing values
• Not all data is correctly filled in by the users. Like year of registration in year 1000 or 9999, powerPS 20000.

# Cleaning the dataset

In [8]:
data.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [9]:
#cleaning column names1
data.columns = data.columns.str.strip().str.lower()
data.columns

Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'yearofregistration', 'gearbox', 'powerps', 'model',
       'kilometer', 'monthofregistration', 'fueltype', 'brand',
       'notrepaireddamage', 'datecreated', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')

In [10]:
#cleaning column names2
data.rename(columns={'datecrawled':'date_crawled',
                     'offertype':'offer_type',
                     'vehicletype': 'vehicle_type',
                    'yearofregistration': 'registration_year',
                    'monthofregistration': 'registration_month',
                    'fueltype': 'fuel_type',
                    'notrepaireddamage': 'unrepaired_damage',
                    'datecreated':'date_created',
                    'nrofpictures': 'nr_of_pictures',
                    'postalcode': 'zip_code',
                    'lastseen' : 'last_seen'}, inplace=True)
data.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'nr_of_pictures', 'zip_code',
       'last_seen'],
      dtype='object')

In [11]:
#Check 'seller' column
data.seller.value_counts(dropna=False)

privat        371525
gewerblich         3
Name: seller, dtype: int64

In [12]:
#The info in the seller column is not usefull to us. Let's remove this column:
data = data.drop('seller', axis=1)
data.head(2)

date_crawled                  name offer_type  price abtest  \
0  2016-03-24 11:52:17            Golf_3_1.6    Angebot    480   test   
1  2016-03-24 10:58:45  A5_Sportback_2.7_Tdi    Angebot  18300   test   

  vehicle_type  registration_year  gearbox  powerps model  kilometer  \
0          NaN               1993  manuell        0  golf     150000   
1        coupe               2011  manuell      190   NaN     125000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi                ja   

          date_created  nr_of_pictures  zip_code            last_seen  
0  2016-03-24 00:00:00               0     70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0     66954  2016-04-07 01:46:50

In [13]:
#Check 'offer_type' column
data.offer_type.value_counts(dropna=False)

Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64

In [14]:
# We do not need the cars listed with offer type 'Gesuch'. Let's remove those rows 
data = data[data['offer_type'] != 'Gesuch']

#and then remove the column offer_type
data = data.drop('offer_type', axis=1)

In [15]:
#Check 'price' column
data.price.value_counts(dropna=False).sort_index()

0             10772
1              1189
2                12
3                 8
4                 1
5                26
7                 3
8                 9
9                 8
10               84
11                5
12                8
13                7
14                5
15               27
16                2
17                5
18                3
19                3
20               51
21                1
24                1
25               33
26                1
27                1
29                2
30               55
32                1
33                1
35               18
              ...  
745000            2
780000            1
820000            1
849000            1
911911            1
999990            1
999999           13
1000000           5
1010010           1
1111111           2
1234566           1
1250000           2
1300000           1
1600000           2
2795000           1
2995000           1
3890000           1
3895000           1
9999999           3


In [16]:
#Check 'price' column1
data.price.value_counts(dropna=False).sort_index()

0             10772
1              1189
2                12
3                 8
4                 1
5                26
7                 3
8                 9
9                 8
10               84
11                5
12                8
13                7
14                5
15               27
16                2
17                5
18                3
19                3
20               51
21                1
24                1
25               33
26                1
27                1
29                2
30               55
32                1
33                1
35               18
              ...  
745000            2
780000            1
820000            1
849000            1
911911            1
999990            1
999999           13
1000000           5
1010010           1
1111111           2
1234566           1
1250000           2
1300000           1
1600000           2
2795000           1
2995000           1
3890000           1
3895000           1
9999999           3


In [17]:
#Check 'price' column2
data[data.price >400000]

date_crawled  \
1846    2016-03-10 22:47:05   
10649   2016-04-05 09:36:18   
14663   2016-04-03 06:03:56   
16889   2016-03-15 16:45:45   
20143   2016-03-18 00:55:53   
20746   2016-03-08 15:54:52   
21467   2016-03-19 19:45:02   
26327   2016-03-21 19:43:54   
38445   2016-03-21 19:53:52   
39625   2016-03-08 23:53:51   
41980   2016-03-11 09:37:17   
45055   2016-04-02 00:06:21   
45833   2016-03-15 10:53:11   
53507   2016-03-11 18:57:57   
54295   2016-03-08 11:51:40   
54507   2016-03-09 11:40:15   
56973   2016-03-18 18:52:27   
57594   2016-03-25 09:56:28   
69726   2016-03-17 15:58:57   
69747   2016-03-20 10:59:32   
77185   2016-03-21 19:50:05   
77520   2016-03-22 14:58:27   
78265   2016-03-31 19:55:47   
78480   2016-03-10 20:43:12   
85047   2016-03-28 12:45:00   
87799   2016-03-08 20:39:05   
90853   2016-04-01 05:36:16   
91106   2016-03-30 19:46:05   
100851  2016-03-12 18:46:35   
106651  2016-03-11 22:54:08   
...                     ...   
281833  2016-03-19 08:50:50   
287835  2016-03-26 00:56:36   
299872  2016-03-14 10:06:23   
301198  2016-04-04 09:36:42   
306469  2016-03-17 12:06:35   
310924  2016-03-14 05:36:22   
312109  2016-04-05 11:50:18   
312909  2016-03-17 09:53:08   
316200  2016-04-04 11:53:31   
318143  2016-03-21 13:57:51   
325589  2016-03-27 21:55:14   
330117  2016-04-01 15:56:24   
332028  2016-03-31 18:39:53   
332531  2016-03-20 02:36:23   
333699  2016-03-12 03:36:18   
337851  2016-03-24 18:59:49   
338415  2016-03-26 22:49:39   
344196  2016-03-05 17:25:25   
347435  2016-03-30 15:50:46   
358868  2016-03-16 16:49:53   
359083  2016-03-29 11:38:54   
360768  2016-03-21 17:39:22   
361451  2016-04-03 14:36:47   
362782  2016-03-06 10:38:42   
363434  2016-03-23 06:36:20   
364171  2016-04-04 21:25:21   
365461  2016-03-31 18:53:48   
366653  2016-03-20 10:59:52   
366861  2016-03-21 10:53:06   
367041  2016-03-24 19:00:03   

                                                     name     price   abtest  \
1846    BMW_M1_Museumsfahrzeug_Neuwagenzustand_Glossy_...    579000  control   
10649                       Porsche_996_GT2_MK2_Clubsport    420000  control   
14663                          Tausch_moeglich_Polo_vw_n9  11111111  control   
16889                    Ford_Mondeo_Kombi_TÜV_abgelaufen   1000000  control   
20143                          Porsche_911R_Einer_von_911   1250000     test   
20746                                  Golf3_Schlachtfest  12345678     test   
21467   Porsche_911_911R_1_of_20_ORIGINAL_R_nur_798KG!...   9999999  control   
26327                                       Porsche_911_R    600000     test   
38445                                       2_VW_Busse_T3    999999     test   
39625                         Tausche_volvo_v40_gegen_van  12345678  control   
41980                                       Maserati_MC12   2995000  control   
45055   Ferrari_LaFerrari_Deutsche_Auslieferung._A_Cel...   2795000  control   
45833                             Porsche_911_Turbo__930_   1000000  control   
53507          Suzuki_SJ_Samurai_413_Van__Kein_Lada_Niva_   1000000     test   
54295                        Mazda_5_2.0_CD_DPF_Exclusive  10000000  control   
54507                    Tausch_DKW_F8_gegen_Lanz_Bulldog  11111111     test   
56973     tausche_ford_mondeo_mk3_ghia_gegen_anderes_auto  99999999  control   
57594                                       Astra_f_teile   1111111     test   
69726                                        Oldtimer_Lkw    999999     test   
69747                    Suche_Mercedes_Benz_r_klasse_cdi  99999999  control   
77185                            Aston_Martin_DB_7_Zagato    500000     test   
77520                         Tausch_gegen_gleichwertiges  99999999  control   
78265   McLaren_675LT_Spider/Delivery_on1._May/Vat_is_...    820000  control   
78480                                       Porsche_991_R   1250000     test   
85047                                         Porsche_911    911911     test   
87799       

In [18]:
#Change prices over 600.000 to NaN
data.loc[data['price'] >600000, 'price'] = np.nan

In [19]:
#Check 'abtest' column
data['abtest'].value_counts(dropna=False)

test       192581
control    178935
Name: abtest, dtype: int64

In [20]:
#Check 'vehicle_type' column
data['vehicle_type'].value_counts(dropna=False)

limousine     95894
kleinwagen    80021
kombi         67564
NaN           37862
bus           30199
cabrio        22898
coupe         19014
suv           14707
andere         3357
Name: vehicle_type, dtype: int64

In [21]:
#Check 'registration_year' column
data['registration_year'].value_counts(dropna=False).sort_index()

1000     38
1001      1
1039      1
1111      4
1200      1
1234      4
1253      1
1255      1
1300      2
1400      1
1500      5
1600      2
1602      1
1688      1
1800      5
1910    104
1911      1
1915      1
1919      1
1920      1
1923      3
1925      1
1927      2
1928      2
1929      8
1930      5
1931      4
1932      4
1933      6
1934      5
       ... 
3700      1
3800      1
4000      3
4100      1
4500      2
4800      1
5000     18
5300      1
5555      2
5600      1
5900      1
5911      2
6000      6
6200      1
6500      1
7000      4
7100      1
7500      2
7777      1
7800      1
8000      2
8200      1
8455      1
8500      1
8888      2
9000      5
9229      1
9450      1
9996      1
9999     27
Name: registration_year, Length: 155, dtype: int64

In [22]:
# Update registration year >2018 and <1801 to NaN
data.loc[data["registration_year"]>2018, "registration_year"] = np.nan
data.loc[data["registration_year"]<1801, "registration_year"] = np.nan

In [23]:
#Check 'gearbox' column
data['gearbox'].value_counts(dropna=False)

manuell      274208
automatik     77105
NaN           20203
Name: gearbox, dtype: int64

In [24]:
#Check 'powerps' column
data['powerps'].value_counts(dropna=False).sort_index()

0        40812
1           34
2           10
3            9
4           30
5          103
6           11
7           11
8            7
9            5
10          17
11          28
12          24
13           8
14          21
15          21
16          11
17           8
18          54
19          15
20          23
21           6
22           5
23          45
24          31
25          16
26         291
27          40
28           6
29          35
         ...  
11509        1
11530        1
11620        1
11635        1
12012        1
12510        2
12512        3
12684        1
13616        1
13636        1
14009        1
15001        1
15016        1
15017        1
15020        1
15033        1
16011        1
16051        1
16311        1
16312        1
17011        1
17019        1
17322        1
17410        1
17700        1
17932        1
19208        1
19211        1
19312        1
20000        1
Name: powerps, Length: 794, dtype: int64

In [25]:
data.sort_values(by='powerps', ascending=False) 

date_crawled  \
230298  2016-03-09 15:43:53   
313723  2016-03-09 08:53:11   
119715  2016-04-03 07:36:21   
138950  2016-03-25 14:55:49   
67100   2016-03-28 11:49:56   
206116  2016-03-20 16:51:48   
158057  2016-03-09 18:57:50   
55501   2016-03-22 20:48:06   
196469  2016-03-17 19:50:23   
183992  2016-03-19 11:38:23   
227574  2016-03-10 08:57:25   
326527  2016-04-02 15:50:55   
190146  2016-03-17 19:39:39   
260714  2016-03-22 22:55:30   
215849  2016-03-12 16:40:34   
155718  2016-03-23 15:55:50   
273571  2016-03-11 13:59:35   
89370   2016-03-18 00:51:03   
170136  2016-04-04 14:59:24   
284182  2016-03-17 23:53:06   
238730  2016-03-30 16:38:22   
224610  2016-03-11 19:44:07   
197143  2016-04-04 15:53:25   
266186  2016-03-17 16:55:13   
306393  2016-03-21 12:49:22   
363454  2016-03-25 12:55:36   
81570   2016-03-25 13:36:57   
171721  2016-03-28 15:53:47   
66039   2016-03-27 18:47:59   
105618  2016-03-22 16:58:28   
...                     ...   
200712  2016-03-26 11:55:37   
200704  2016-03-25 12:54:01   
333798  2016-03-18 16:37:21   
200701  2016-03-31 20:36:19   
200697  2016-04-05 12:57:47   
200855  2016-04-02 17:57:15   
200869  2016-03-07 10:57:33   
201033  2016-03-19 17:48:37   
200952  2016-03-26 11:50:44   
201026  2016-03-21 22:47:45   
333724  2016-04-06 16:06:19   
200998  2016-03-12 20:54:31   
23385   2016-03-16 17:46:02   
200981  2016-03-17 15:51:29   
200974  2016-03-19 19:54:17   
333734  2016-03-11 11:06:25   
5228    2016-03-12 07:58:56   
200957  2016-03-22 19:50:56   
200950  2016-03-30 10:57:23   
200872  2016-04-01 19:06:23   
5225    2016-03-10 18:47:39   
200927  2016-03-21 21:54:58   
333750  2016-03-11 11:37:36   
333751  2016-03-26 12:49:16   
333752  2016-03-10 15:58:04   
5224    2016-03-12 13:52:18   
23399   2016-03-30 22:38:41   
200888  2016-04-04 13:54:02   
200881  2016-03-14 21:42:55   
0       2016-03-24 11:52:17   

                                                     name    price   abtest  \
230298                     Mercedes_CLK_430___8_Zylinder_   4300.0  control   
313723                                  Verkaufe_BMW_528i   1500.0     test   
119715            BMW_120D_Diesel_M_packet_Vollauslastung   9999.0     test   
138950                                   BMW_E39_5er_525i   2100.0  control   
67100                                          Opel_Omega   3250.0  control   
206116                           Verkaufe_meinen_bmw_525d   6000.0     test   
158057               Audi_80_Avant_2.8_V6_RS2_Optik_/_LPG   1700.0     test   
55501                        BMW_E70_X5_3.0D_Panoramadach  23333.0  control   
196469  BMW_E30_touring_325i_Projektaufgabe_325e_viele...   1749.0  control   
183992              Audi_a3_2_0_tdi_s_line_plus_125kW_dsg  10900.0     test   
227574                               Top_Cabrio_A4_s_line  11950.0     test   
326527  BMW_520_Touring__mit__Austauschmoter_38tsd_km_...  11900.0  control   
190146  Mercedes_Benz_Oldtimer_W108_250SE_Automatik_Gu...   5250.0  control   
260714                            Skoda_Octavia_Combi_Tsi   5000.0  control   
215849                                      Alfa_Romeo_GT   3150.0  control   
155718                                     Alfa_Romeo_147   3650.0     test   
273571                                           Bmw_320d   1750.0     test   
89370   PT_Cruiser_Crd_2.0_fuer_Kurzentschlossene_bis_...      0.0     test   
170136                          Top_geopflegter_fiesta_st   5200.0     test   
284182                                       Opel_Astra_j   9799.0  control   
238730  Mercedes_C_200_CDI_BE_T_Modell_Silber_TOP_GEPF...  15499.0     test   
224610                            Mercedes_E_klasse_combi    500.0     test   
197143                            Ibiza_Seat_zu_verkaufen    200.0     test   
266186                    Schlachte_Opel_vectra_b_Caravan    300.0     test   
306393                                        Opel_safira   2000.0  control   
363454                                

In [26]:
data[data['powerps']>800]

date_crawled  \
1898    2016-03-22 20:52:00   
2199    2016-03-21 11:55:22   
4254    2016-04-03 20:31:00   
5582    2016-03-29 19:44:48   
8038    2016-04-02 19:25:25   
8099    2016-04-02 11:48:54   
10148   2016-03-14 20:06:19   
10368   2016-03-31 20:51:13   
11577   2016-03-25 19:55:32   
13358   2016-03-09 18:43:44   
16193   2016-03-15 21:53:29   
20008   2016-03-20 11:00:08   
20022   2016-03-20 17:54:04   
20783   2016-03-18 23:52:30   
21020   2016-03-26 19:46:06   
21261   2016-03-25 18:48:26   
22091   2016-03-16 16:46:57   
22650   2016-03-29 17:53:17   
23362   2016-03-17 05:03:48   
25766   2016-03-25 14:55:04   
26429   2016-03-28 19:57:39   
30309   2016-03-21 21:58:07   
31090   2016-03-20 22:55:15   
34342   2016-03-22 14:58:27   
35563   2016-03-09 11:37:03   
35964   2016-04-01 06:02:10   
35992   2016-03-25 18:56:53   
36848   2016-03-16 00:18:43   
37115   2016-03-12 16:46:14   
37924   2016-04-01 01:56:06   
...                     ...   
338501  2016-03-12 17:44:12   
339626  2016-03-11 17:48:48   
339814  2016-04-05 10:37:13   
343150  2016-03-30 19:58:09   
344841  2016-03-20 09:57:28   
348031  2016-03-12 10:56:09   
351721  2016-03-27 01:00:18   
354482  2016-03-07 17:49:02   
354518  2016-04-02 16:57:37   
355280  2016-03-24 19:49:36   
356131  2016-03-07 16:58:44   
357294  2016-04-04 23:55:47   
357617  2016-04-03 22:57:04   
357868  2016-03-07 16:42:25   
357890  2016-03-29 21:40:33   
358921  2016-03-22 09:57:36   
362535  2016-03-30 17:56:27   
363454  2016-03-25 12:55:36   
363507  2016-03-24 11:50:41   
363852  2016-03-21 20:58:26   
364886  2016-03-22 13:44:40   
365025  2016-03-16 21:48:39   
365133  2016-04-04 12:38:11   
365581  2016-03-05 17:53:09   
365751  2016-03-22 00:38:15   
365806  2016-03-13 11:51:57   
365894  2016-04-04 18:53:25   
369007  2016-03-07 21:36:19   
370612  2016-04-02 20:54:21   
371414  2016-03-09 17:59:03   

                                                     name    price   abtest  \
1898                                       Toyota_Corolla   3200.0  control   
2199                Reparaturbeduerfdiger_Opel_Vectra_1.8      0.0  control   
4254                           Mitsubishi_Colt_Diesel_1_5   3100.0  control   
5582                             verkauf__bastler__export    500.0  control   
8038    VW_LUPO_1_4_16V_SCHWARZ_75PS_SPORTFAHRWERK_ABT...   1499.0     test   
8099                                          Honda_Civic   1500.0     test   
10148                                   Opel_zafira_2999€   2599.0  control   
10368                     Ford_Focus_Kombi_Schwarz_DEFEKT    550.0     test   
11577                       Fiat_Lancia_Delta_HPE_1_6_16v      0.0     test   
13358                                Ford_Streetka_1.6_8V   3500.0     test   
16193                         Ford_Escort_RS_Cosworth_NEU      0.0  control   
20008                                  Seat_ibiza_1.4_MPI      0.0  control   
20022                                     BMW_535_touring   9750.0     test   
20783   Nissan_Micra___treues_Rentnerauto_&_Schmuckstueck   3390.0     test   
21020                                 Mercedes_Benz_e_270   2200.0  control   
21261   CITROEN_C5_Diesel_HDI__Gute_Zustand_Gruene_Pla...   2000.0     test   
22091              Opel_Corsa_D_1_2_Cosmo_Teilleder_Klima   3990.0  control   
22650                                    lancia_y_schwarz    200.0  control   
23362                                   BMW_328i_schalter   3999.0  control   
25766                             1.HAND__wenig_Kilometer   1499.0  control   
26429                               VW_CADDY_MAXI_1.9_TDI  10900.0  control   
30309                Vw_Golf_3_Varriant__KEINE_PAPIERE_!!    300.0  control   
31090                      Volvo_v40_Diesel__silber_Kombi   1199.0     test   
34342                       Chevrolet_g20_mit_nur_39000km   9500.0  control   
35563       Ideal_als_Zweitwagen_oder_fuer_Fahranfaenger.   3740.0     test   
35964                                 

In [27]:
#update PS >800 and PS <30 to NaN
data.loc[data['powerps']>800, 'powerps'] = np.nan
data.loc[data['powerps'] < 30, 'powerps'] = np.nan

In [28]:
data.sort_values(by='powerps', ascending=True) 

date_crawled  \
118540  2016-03-05 16:44:57   
242405  2016-03-27 12:36:22   
66498   2016-03-17 22:06:21   
242236  2016-04-04 21:54:38   
244457  2016-04-05 00:36:45   
308802  2016-03-07 12:53:15   
246889  2016-03-08 20:49:38   
325075  2016-03-07 17:48:49   
161967  2016-03-09 01:47:48   
65304   2016-04-04 19:37:42   
327347  2016-03-08 14:52:13   
40833   2016-03-20 11:59:21   
240839  2016-03-31 22:50:46   
18677   2016-03-19 17:53:13   
36567   2016-03-29 17:58:45   
352061  2016-03-15 00:36:35   
363634  2016-04-04 20:46:38   
307314  2016-03-21 19:50:09   
16174   2016-03-30 23:38:13   
131387  2016-03-05 19:52:58   
263187  2016-03-11 14:46:25   
350965  2016-03-18 21:37:22   
3491    2016-03-15 21:36:20   
7402    2016-03-11 09:53:53   
240034  2016-03-23 15:49:47   
309833  2016-03-17 17:45:07   
359741  2016-04-04 20:55:58   
52649   2016-04-05 08:25:24   
23714   2016-03-23 11:37:57   
168626  2016-03-16 07:54:19   
...                     ...   
371296  2016-03-12 13:57:47   
371299  2016-03-25 12:54:45   
371309  2016-03-07 19:51:14   
371316  2016-03-29 19:50:03   
371342  2016-04-01 01:56:04   
371348  2016-04-05 10:57:00   
371349  2016-03-10 23:36:19   
371351  2016-03-30 23:48:01   
371359  2016-03-26 08:54:05   
371370  2016-03-19 14:59:31   
371374  2016-03-14 23:53:14   
371375  2016-03-10 19:36:59   
371380  2016-03-14 11:51:40   
371407  2016-03-25 09:37:59   
371414  2016-03-09 17:59:03   
371417  2016-03-14 23:00:21   
371424  2016-04-04 13:55:18   
371443  2016-03-16 13:36:25   
371457  2016-03-12 08:37:15   
371460  2016-04-03 13:46:24   
371473  2016-03-15 19:57:11   
371489  2016-03-13 15:45:22   
371492  2016-04-04 14:41:28   
371493  2016-03-05 18:57:45   
371495  2016-03-31 19:52:33   
371504  2016-03-07 17:06:35   
371518  2016-04-02 20:37:03   
371521  2016-03-27 20:36:20   
371522  2016-03-21 09:50:58   
371523  2016-03-14 17:48:27   

                                                     name    price   abtest  \
118540                                        NSU_Prinz_4   2000.0  control   
242405                               Nissan_Micra_k11_1_0    200.0     test   
66498                                      Yamaha_warrior   3750.0  control   
242236                                        NSU_Prinz_4   1950.0     test   
244457                            Volkswagen_ovali_export   6900.0  control   
308802          Daihatsu_CUORE_"_SELTENER_OLDTIMER_"_LOOK   3500.0     test   
246889                                   Smart_Fortwo_CDI   2300.0  control   
325075                               VW_Kaefer_Ovali_1954   6600.0     test   
161967  VW_Kaefer__&#x27;62__Dickholmer__Faltdach__Exp...   3999.0  control   
65304                             Fiat_Panda_30___1_Serie    600.0     test   
327347                                   Smart_fortwo_CDI   2300.0     test   
40833                                   Volkswagen_Kaefer  18500.0     test   
240839                                        NSU_Prinz_4   5400.0     test   
18677                                   unimog_411_cabrio  17900.0  control   
36567   OPEL_CORSA_C_MIT_KLIMA_UND_SITZHEIZUNG_WENIG_K...   1250.0  control   
352061                     Oldtimer_Aero_30_Roadster_1938  17000.0     test   
363634                                     Citroën_Andere   9850.0  control   
307314                                  Volkswagen_Kaefer   8750.0     test   
16174                                        Kaefer_Ovali  18000.0  control   
131387       Kafer_1957_1958_Chassis/Bodengruppe_komplett   1995.0     test   
263187                 Komplettlackierung_fuer_Kleinwagen      0.0  control   
350965  VW_Winker_Faltdach_Kaefer_Bj_1958_zum_Fertigst...   3800.0  control   
3491     Ford_Popular_Oldtimer_103_E_Restauriert_Oldtimer  10500.0  control   
7402                                         Smart_Diesel   1300.0  control   
240034                             NSU_Prinz_4_L__Typ_47_   1550.0     test   
309833                                

Still in 'powerps' lots of unrealistic values.

In [29]:
#Check 'model' column
data['model'].value_counts(dropna=False)

golf                  30069
andere                26399
3er                   20566
NaN                   20481
polo                  13092
corsa                 12573
astra                 10829
passat                10306
a4                    10257
c_klasse               8775
5er                    8546
e_klasse               7560
a3                     6604
a6                     6023
focus                  5950
fiesta                 5774
transporter            5527
twingo                 4953
2_reihe                4816
fortwo                 4338
a_klasse               4316
vectra                 4255
1er                    3837
mondeo                 3626
clio                   3552
touran                 3540
3_reihe                3503
punto                  3349
zafira                 3049
megane                 2927
                      ...  
mii                      81
crossfire                66
range_rover_evoque       65
gl                       64
nubira              

In [30]:
#Check 'kilometer' column
data['kilometer'].value_counts(dropna=False)

150000    240790
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7614
5000        7066
40000       6376
30000       6041
20000       5675
10000       1949
Name: kilometer, dtype: int64

In [31]:
#Check 'fuel_type' column
data['fuel_type'].value_counts(dropna=False)

benzin     223854
diesel     107744
NaN         33379
lpg          5378
cng           571
hybrid        278
andere        208
elektro       104
Name: fuel_type, dtype: int64

In [32]:
#Check 'brand' column
data['brand'].value_counts(dropna=False)

volkswagen        79638
bmw               40272
opel              40135
mercedes_benz     35308
audi              32873
ford              25571
renault           17968
peugeot           11027
fiat               9676
seat               7022
mazda              5695
skoda              5641
smart              5249
citroen            5182
nissan             5037
toyota             4694
sonstige_autos     3980
hyundai            3646
mini               3394
volvo              3327
mitsubishi         3061
honda              2836
kia                2555
alfa_romeo         2345
suzuki             2328
porsche            2215
chevrolet          1845
chrysler           1452
dacia               900
jeep                807
daihatsu            806
subaru              779
land_rover          770
jaguar              621
trabant             590
daewoo              542
saab                530
rover               490
lancia              484
lada                225
Name: brand, dtype: int64

In [33]:
#Check 'nr_of_pictures' column
data['nr_of_pictures'].value_counts(dropna=False)

0    371516
Name: nr_of_pictures, dtype: int64

In [34]:
#Remove 'nr_of_pictures' column
data = data.drop('nr_of_pictures', axis=1)

In [35]:
#Check 'zip_code' column
data['zip_code'].value_counts(dropna=False)

10115    828
65428    637
66333    349
38518    326
44145    323
32257    323
52525    314
78224    309
26789    301
48599    294
13357    293
53757    284
50354    283
48249    282
40764    282
60311    281
51065    278
52249    277
60386    275
61169    269
92637    266
26871    265
66424    265
45881    262
56070    262
65719    261
13409    257
87700    257
59192    254
44339    254
        ... 
55499      1
67311      1
21784      1
38325      1
63874      1
21360      1
95356      1
26906      1
29361      1
17154      1
79685      1
91471      1
88376      1
77787      1
95694      1
27254      1
60306      1
7619       1
29587      1
91465      1
25946      1
4685       1
98749      1
74937      1
24638      1
97859      1
29484      1
85071      1
17329      1
37297      1
Name: zip_code, Length: 8150, dtype: int64

In [36]:
#change date+time columns to date only
data["date_crawled"] = data["date_crawled"].str[:10]
data["date_created"] = data["date_created"].str[:10]
data["last_seen"] = data["last_seen"].str[:10]

In [37]:
#Update model values
data['brand'] = data['brand'].str.capitalize()

In [38]:
data.head()

date_crawled                            name    price abtest vehicle_type  \
0   2016-03-24                      Golf_3_1.6    480.0   test          NaN   
1   2016-03-24            A5_Sportback_2.7_Tdi  18300.0   test        coupe   
2   2016-03-14  Jeep_Grand_Cherokee_"Overland"   9800.0   test          suv   
3   2016-03-17              GOLF_4_1_4__3TÜRER   1500.0   test   kleinwagen   
4   2016-03-31  Skoda_Fabia_1.4_TDI_PD_Classic   3600.0   test   kleinwagen   

   registration_year    gearbox  powerps  model  kilometer  \
0             1993.0    manuell      NaN   golf     150000   
1             2011.0    manuell    190.0    NaN     125000   
2             2004.0  automatik    163.0  grand     125000   
3             2001.0    manuell     75.0   golf     150000   
4             2008.0    manuell     69.0  fabia      90000   

   registration_month fuel_type       brand unrepaired_damage date_created  \
0                   0    benzin  Volkswagen               NaN   2016-03-24   
1                   5    diesel        Audi                ja   2016-03-24   
2                   8    diesel        Jeep               NaN   2016-03-14   
3                   6    benzin  Volkswagen              nein   2016-03-17   
4                   7    diesel       Skoda              nein   2016-03-31   

   zip_code   last_seen  
0     70435  2016-04-07  
1     66954  2016-04-07  
2     90480  2016-04-05  
3     91074  2016-03-17  
4     60437  2016-04-06

In [39]:
data.describe(include='all')

date_crawled         name          price  abtest vehicle_type  \
count        371516       371516  371422.000000  371516       333654   
unique           34       233520            NaN       2            8   
top      2016-04-03  Ford_Fiesta            NaN    test    limousine   
freq          14390          657            NaN  192581        95894   
mean            NaN          NaN    5754.180409     NaN          NaN   
std             NaN          NaN    9496.782999     NaN          NaN   
min             NaN          NaN       0.000000     NaN          NaN   
25%             NaN          NaN    1150.000000     NaN          NaN   
50%             NaN          NaN    2950.000000     NaN          NaN   
75%             NaN          NaN    7200.000000     NaN          NaN   
max             NaN          NaN  600000.000000     NaN          NaN   

        registration_year  gearbox        powerps   model      kilometer  \
count       371307.000000   351313  329432.000000  351035  371516.000000   
unique                NaN        2            NaN     251            NaN   
top                   NaN  manuell            NaN    golf            NaN   
freq                  NaN   274208            NaN   30069            NaN   
mean          2003.347478      NaN     126.200648     NaN  125619.690673   
std              7.776021      NaN      62.161008     NaN   40110.814656   
min           1910.000000      NaN      30.000000     NaN    5000.000000   
25%           1999.000000      NaN      80.000000     NaN  125000.000000   
50%           2003.000000      NaN     116.000000     NaN  150000.000000   
75%           2008.000000      NaN     150.000000     NaN  150000.000000   
max           2018.000000      NaN     800.000000     NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count        371516.000000    338137      371516            299462   
unique                 NaN         7          40                 2   
top                    NaN    benzin  Volkswagen              nein   
freq                   NaN    223854       79638            263178   
mean              5.734555       NaN         NaN               NaN   
std               3.712372       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

       date_created       zip_code   last_seen  
count        371516  371516.000000      371516  
unique          114            NaN          34  
top      2016-04-03            NaN  2016-04-06  
freq          14449            NaN       80930  
mean            NaN   50820.754880         NaN  
std             NaN   25799.143633         NaN  
min             NaN    1067.000000         NaN  
25%             NaN   30459.000000         NaN  
50%             NaN   49610.000000         NaN  
75%             NaN   71546.000000         NaN  
max             NaN   99998.000000         NaN

In [40]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 371516 entries, 0 to 371527
Data columns (total 17 columns):
date_crawled          371516 non-null object
name                  371516 non-null object
price                 371422 non-null float64
abtest                371516 non-null object
vehicle_type          333654 non-null object
registration_year     371307 non-null float64
gearbox               351313 non-null object
powerps               329432 non-null float64
model                 351035 non-null object
kilometer             371516 non-null int64
registration_month    371516 non-null int64
fuel_type             338137 non-null object
brand                 371516 non-null object
unrepaired_damage     299462 non-null object
date_created          371516 non-null object
zip_code              371516 non-null int64
last_seen             371516 non-null object
dtypes: float64(3), int64(3), object(11)
memory usage: 51.0+ MB


## Saving the cleaned dataset

In [42]:
#Save the cleaned dataset
data.to_csv('Data/autos_cleaned.csv', index=False)